### ISSUE TITLE
Cannot find good ways to test Created AtLocation

### ISSUE THREAD

**Issue by foresightyj** [ivaylokenov/MyTested.WebApi#247]
I have a post action that returns 201 Created with a Location header and an empty body. Basically I want to write my test like this:

    .ShouldReturn()
    .Created()
    .AtLocation(loc => loc.WithAbsolutePath("/api/v1/friends/_outgoing/%d+"));

I don't care about the actual value of the ID (%d) there, as long as it is a number. I wish I could write a regular expression string like that. But this is not possible. The `.At<FriendController>` method also requires explicitly supplying an ID.

Or, I wish I could write something like this:

    .AtLocation(loc =>
    {
        string location = loc.Value;
        Assert.IsTrue(location.ToLower().StartsWith("/api/v1/friends/_outgoing/"));
    })

I also tried to get hold of a reference to the response message so I could manually assert myself but I could find a method called `AndProvideTheHttpResponseMessage` but it wasn't there.

![image](https://cloud.githubusercontent.com/assets/1296736/12370265/6cd32a84-bc45-11e5-8750-f529cfdc36d9.png)

I'd prefer `AndProvideTheHttpResponseMessage` method there so I can read the actual ID and go on testing if the uri specified at Location returns 200.

**Comment by ivaylokenov** [comment]
@foresightyj Will this method work for your case:

```c#
MyWebApi
    .Controller<WebApiController>()
    .Calling(c => c.SomeAction())
    .ShouldReturn()
    .Created()
    .AtLocation("http://somehost.com/someuri/1?query=someQuery");
```

I am currently thinking of a way to test parts of strings everywhere.

About the `HttpResponseMessage` - it will be hard for controller testing because it will require a lot of pipeline mocking and it is not the point of the unit test. This is why integration tests are there. You can use them. I will think of a way to add it to the controller but it may not be the same HTTP response message as the one you will get on calling the API server.

**Comment by foresightyj** [comment]
Thanks. I didn't know all the complexities in `HttpResponseMessage`.

`AtLocation` itself doesn't work for me as I cannot provide an exact uri string (since I don't know the ID).

I don't need the complete `HttpResponseMessage`, I only need a way to read uri in the Location header in a Created 201 Response. Is there a way to read it in the current version?

I already used the Integration Test in MyTested.WebApi. Will resort to that if unit testing it is impossible.

**Comment by ivaylokenov** [comment]
@foresightyj  Integration test is the way here. Otherwise, you need to provided a mocked ID, which will come later in the string URL. Or am I missing something? Can you provide some sort of code sample to see the case you are trying to test?

**Comment by foresightyj** [comment]
[MyTestedWebApiCreated201.zip](https://github.com/ivaylokenov/MyTested.WebApi/files/94432/MyTestedWebApiCreated201.zip)

I provided a minimal test case, which is closet to my actual project.

I realized that my unit test might deviated from a typical unit test a bit. 

My project models a social network which has users, friendships, posts, comments, etc. I started without the persistent layer so I could concentrate on the business layer + presentation layer first. For demo purpose, I set up a large graph of interconnected domain objects in memory, in which a few users and their friendships exist beforehand. In Application_Start, I used Ninject to inject a singleton copy of these objects to controllers so state is maintained across requests ( ignore thread safety) until the server restarts.

I have a typical layered architecture: controller --> service layer --> repositories. Repositories read from those in memory objects directly. For every test case, I will initialize a new copy of the graph of objects, **instead of mocking just enough data for that test**.

Back to the 201 Created example, when I created something, the newly created objects is inserted into a collection inside the graph of in memory objects. Unless I dig into these objects, I would not find the newly created ID.

**Sorry for such a long reply. It is OK if it does not make sense to you. I also had a hard time to explain it.**

**Comment by ivaylokenov** [comment]
@foresightyj I believe you need to mock your database object, which the controller is using. Then set it to always apply constant ID when adding object and then test with it. If this does not work for you, I will implement string tester with StartsWith for example.

**Comment by foresightyj** [comment]
I already mocked the database (which is what I mentioned as the graph of interconnected domain objects in memory). This mocked database together with repositories that access it is both used for **unit test/integration test** as well as the **actual application for demo purpose**. So the repository is not mocked ad hoc but is one that I've written beforehand that is used for both test and actual deployment.

Probably very few people has done things this way so my usage is not very generic. Feel free to ignore my use case. Thanks.

**Comment by ivaylokenov** [comment]
@foresightyj I will add string test builder which will allow to test all strings with StartsWith, EndsWith and so on, so it will work for your case. Thank you for using the library and making sugestions. :+1:

**Comment by foresightyj** [comment]
@ivaylokenov I really appreciate your kindness in supporting that. 

I absolutely love this library and it greatly improved the unit/integration test experience in my team. I learned a lot about Web API and OWIN's internals while using this library.

Keep up the good work and I will read the source code when I am less busy and hopefully can contribute via pull requests some day.

**Comment by ivaylokenov** [comment]
@foresightyj You are welcome and thank you for the kind words. I am glad the library helps peope write tests faster and easier. More updates/libraries are comming.

**Comment by ivaylokenov** [comment]
@foresightyj Can I put your comment as a testimonial on the project's web site - http://mytestedasp.net/ ?
If yes, can I have your name, job title or whatever information you are willing to provide. Thank you!

**Comment by foresightyj** [comment]
No problem. I work for a tiny startup in China:

Name: Yuan Jian
Company: Jiangsu HuoHuo Network Technology Co., Ltd. (Nanjing, China)
Job title: Project Director
The number of people in the dev team: 5

**Comment by ivaylokenov** [comment]
@foresightyj Implemented in version 1.2.1. You can update the package. There is another method AtLocationPassing which allows predicate assertions.

Thank you for the information. Will update the site soon.
